src/api/cmd/chart-repo-sync/sync.go

```diff
@@ -0,0 +1,335 @@
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/disintegration/imaging"
+	"github.com/ghodss/yaml"
+	"github.com/jinzhu/copier"
+	"github.com/kubeapps/common/datastore"
+	"gopkg.in/mgo.v2/bson"
+	helmrepo "k8s.io/helm/pkg/repo"
+)
+
+const (
+	chartCollection       = "charts"
+	chartReadmeCollection = "readmes"
+)
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+var dbSession datastore.Session
+
+var netClient httpClient = &http.Client{
+	Timeout: time.Second * 10,
+}
+
+func main() {
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stderr, "Usage: %s [options] [REPO NAME] [REPO URL]\n", os.Args[0])
+		flag.PrintDefaults()
+	}
+	dbURL := flag.String("mongo-url", "localhost", "MongoDB URL (see https://godoc.org/labix.org/v2/mgo#Dial for format)")
+	dbName := flag.String("mongo-database", "charts", "MongoDB database")
+	dbUsername := flag.String("mongo-user", "", "MongoDB user")
+	dbPassword := os.Getenv("MONGO_PASSWORD")
+	debug := flag.Bool("debug", false, "verbose logging")
+	flag.Parse()
+
+	if *debug {
+		log.SetLevel(log.DebugLevel)
+	}
+
+	if flag.NArg() != 2 {
+		flag.Usage()
+		os.Exit(2)
+	}
+
+	mongoConfig := datastore.Config{URL: *dbURL, Database: *dbName, Username: *dbUsername, Password: dbPassword}
+	var err error
+	dbSession, err = datastore.NewSession(mongoConfig)
+	if err != nil {
+		log.WithFields(log.Fields{"host": *dbURL}).Fatal(err)
+	}
+
+	if err := sync(flag.Arg(0), flag.Arg(1)); err != nil {
+		log.WithError(err).Error("sync failed")
+		os.Exit(1)
+	}
+}
+
+// Syncing is performed in the following sequential steps:
+// 1. Update database to match chart metadata from index
+// 3. Process icons for charts
+// 4. Process READMEs for each chart version
+//
+// These steps are processed in this way to ensure relevant chart data is
+// imported into the database as fast as possible. E.g. we want all icons for
+// charts before fetching readmes for each chart and version pair.
```

Comment:
further explanation: this will enable the UI (e.g. Dashboard) to show all charts as soon as possible, but not have icons and readmes for everything right away. The idea is to have Dashboard startup as quickly as possible.

Reply:
As I said in my previous comment, in my opinion, I would only expose the charts that have been fully imported, but the UI will start showing charts as soon as 1 chart has been fully processed, which is not the case today.